### PR TITLE
Implement support for memory (and grow/size/load/store)

### PR DESCRIPTION
**Pull request by axic**

### CODE REVIEW

File: lib/fizzy/execute.cpp [hunk 1]
```diff
@@ -53,6 +53,20 @@ std::vector<uint64_t> execute(const module& _module, funcidx _function, std::vec
...
             goto end;
+        case instr::drop: {
+            stack.pop();
+            break;
+        }
+        case instr::select: {
+            const auto a = static_cast<uint32_t>(stack.pop());
+            const auto b = static_cast<uint32_t>(stack.pop());
+            const auto condition = static_cast<uint32_t>(stack.pop());
+            if (condition == 0)
```
Comment (axic) [inline lib/fizzy/execute.cpp]: Need to check which one is it. The spec is unclear.

Comment (axic) [reply lib/fizzy/execute.cpp]: Fixed.

File: lib/fizzy/execute.cpp [hunk 2]
```diff
@@ -147,6 +168,59 @@ execution_result execute(Instance& instance, FuncIdx function, std::vector<uint6
...
             break;
         }
+        case Instr::memory_size:
+        {
+            stack.push(static_cast<uint32_t>(instance.memory.size() / pagesize));
+            break;
+        }
+        case Instr::memory_grow:
+        {
+            const auto a = static_cast<uint32_t>(stack.pop());
```
Comment (axic) [inline lib/fizzy/execute.cpp]: Renaming `a` to `delta`.

File: lib/fizzy/execute.cpp [hunk 3]
```diff
@@ -79,7 +96,11 @@ inline T rotr(T lhs, T rhs) noexcept
 
 Instance instantiate(const Module& module)
 {
-    return {module};
+    // FIXME: set pages from proper section
+    constexpr unsigned pages = 1;
+    // NOTE: fill it with zeroes
+    bytes memory(pages * pagesize, 0);
+    return {module, memory};
```
Comment (chfast) [inline lib/fizzy/execute.cpp]:
```suggestion
    return {module, std::move(memory)};
```

File: lib/fizzy/parser_expr.cpp [hunk 4]
```diff
@@ -61,6 +63,20 @@ parser_result<Code> parse_expr(const uint8_t* pos)
             push(code.immediates, imm);
             break;
         }
+
+        case Instr::i32_load:
+        case Instr::i32_store:
+        {
+            // alignment
+            uint32_t imm;
+            std::tie(imm, pos) = leb128u_decode<uint32_t>(pos);
```
Comment (chfast) [inline lib/fizzy/parser_expr.cpp]:
```suggestion
            std::tie(std::ignore, pos) = leb128u_decode<uint32_t>(pos);
```

**Comment by axic** [comment]
Also adding parsing of memory section.

File: lib/fizzy/execute.cpp [hunk 5]
```diff
@@ -6,13 +6,10 @@
...
 namespace fizzy
 {
-struct Instance
-{
-    const Module& module;
-};
-
 namespace
 {
+constexpr unsigned pagesize = 65536;
```
Comment (gumb0) [inline lib/fizzy/execute.cpp]: Maybe camel case for consts?

File: lib/fizzy/execute.cpp [hunk 6]
```diff
@@ -129,7 +140,11 @@ inline uint64_t popcnt64(uint64_t value) noexcept
 
 Instance instantiate(const Module& module)
 {
-    return {module};
+    // FIXME: set pages from proper section
+    constexpr unsigned pages = 1;
+    // NOTE: fill it with zeroes
+    bytes memory(pages * pagesize, 0);
```
Comment (gumb0) [inline lib/fizzy/execute.cpp]: Could be
```suggestion
    bytes memory(pages * pagesize);
```

File: lib/fizzy/execute.cpp [hunk 7]
```diff
@@ -129,7 +140,11 @@ inline uint64_t popcnt64(uint64_t value) noexcept
...
+    return {module, std::move(memory)};
```
Comment (gumb0) [inline lib/fizzy/execute.cpp]: I think better without `move` or even just construct inline
```suggestion
    return {module, bytes(pages * pagesize)};
```

File: lib/fizzy/execute.cpp (hunk 5 above)
Comment (axic) [reply lib/fizzy/execute.cpp]: I don't mind. @chfast ?

File: lib/fizzy/execute.cpp (hunk 6 above)
Comment (axic) [reply lib/fizzy/execute.cpp]: Apparently the constructor requires the explicit filling character.

Comment (gumb0) [reply lib/fizzy/execute.cpp]: not for `bytes`, sorry

File: lib/fizzy/execute.cpp (hunk 7 above)
Comment (axic) [reply lib/fizzy/execute.cpp]: I'd prefer to keep the explicit `memory` name for now.

File: lib/fizzy/execute.cpp [hunk 8]
```diff
@@ -205,6 +220,45 @@ execution_result execute(Instance& instance, FuncIdx function, std::vector<uint6
...
+            stack.push(static_cast<uint32_t>(instance.memory.size() / pagesize));
+            break;
+        }
+        case Instr::memory_grow:
+        {
+            const auto delta = static_cast<uint32_t>(stack.pop());
+            const auto cur_pages = instance.memory.size() / pagesize;
+            assert(cur_pages <= size_t(std::numeric_limits<int32_t>::max()));
+            const auto new_pages = cur_pages + delta;
+            assert(cur_pages >= new_pages);
```
Comment (gumb0) [inline lib/fizzy/execute.cpp]: condition looks reversed

File: lib/fizzy/execute.cpp (hunk 5 above)
Comment (chfast) [reply lib/fizzy/execute.cpp]: I don't care.

File: lib/fizzy/execute.cpp [hunk 9]
```diff
@@ -205,6 +222,85 @@ execution_result execute(Instance& instance, FuncIdx function, std::vector<uint6
...
+            store<uint32_t>(instance.memory, address + offset, value);
+            break;
+        }
+        case Instr::i64_store:
+        {
+            const auto address = static_cast<uint32_t>(stack.pop());
+            // NOTE: alignment is dropped by the parser
+            const auto offset = read<uint32_t>(immediates);
+            const auto value = static_cast<uint64_t>(stack.pop());
+            if ((address + offset + sizeof(uint32_t)) > instance.memory.size())
```
Comment (chfast) [inline lib/fizzy/execute.cpp]:
```suggestion
            if ((address + offset + sizeof(uint64_t)) > instance.memory.size())
```

File: lib/fizzy/execute.cpp [hunk 10]
```diff
@@ -205,6 +222,85 @@ execution_result execute(Instance& instance, FuncIdx function, std::vector<uint6
...
+            try
+            {
+                // TODO: remove this once section parsing is done (and enforce limit in parser)
+                if ((new_pages * pagesize) > memorylimit)
+                    throw std::bad_alloc();
+                instance.memory.resize(new_pages * pagesize);
+            }
+            catch (std::bad_alloc const&)
+            {
+                ret = 0xffffffff;  // -1
```
Comment (chfast) [inline lib/fizzy/execute.cpp]:
```suggestion
                ret = static_cast<uint32_t>(-1);
```

**Pull request merged**